### DISCUSSION TITLE
grahpscope可以正常运行，但是在涉及到有动态编译的模块比如pagerank()或者使用GIE时会编译错误

### DISCUSSION THREAD

**Discussion by Zhumeng420** [alibaba/GraphScope#3465] [Q&A] [answered]
<img width="945" alt="截屏2023-12-29 22 24 11" src="https://github.com/alibaba/GraphScope/assets/76473360/62d07c8a-152f-40d6-bbf7-4ec3391e0cc0">
grahpscope可以正常运行，但是在涉及到有动态编译的模块比如pagerank()或者使用GIE时会编译错误。相关组件的版本号也符合官网文档要求。如何解决？

**Comment by Zhumeng420** [comment]
<img width="1299" alt="截屏2023-12-29 22 32 21" src="https://github.com/alibaba/GraphScope/assets/76473360/166fdfdd-4f5d-45b7-a70f-8ad0d6f5b1cf">
附加

**Reply by sighingnow** [reply]
有更完整的报错信息么

**Reply by Zhumeng420** [reply]
2023-12-30 12:23:36,725 [ERROR][rpc:188]: Runstep failed with code: ANALYTICAL_ENGINE_INTERNAL_ERROR, message: Error occurred during RunStep, The traceback is: Traceback (most recent call last):
  File "/Users/zhumeng/anaconda3/envs/graphScopeNew/lib/python3.9/site-packages/gscoordinator/utils.py", line 382, in _compile_on_local
    logger.debug(run_command(command, cwd=workdir))
  File "/Users/zhumeng/anaconda3/envs/graphScopeNew/lib/python3.9/site-packages/gscoordinator/utils.py", line 287, in run_command
    cp = subprocess.run(shlex.split(args), capture_output=True, cwd=cwd)
  File "/Users/zhumeng/anaconda3/envs/graphScopeNew/lib/python3.9/subprocess.py", line 505, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/Users/zhumeng/anaconda3/envs/graphScopeNew/lib/python3.9/subprocess.py", line 951, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/Users/zhumeng/anaconda3/envs/graphScopeNew/lib/python3.9/subprocess.py", line 1837, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'cmake'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/zhumeng/anaconda3/envs/graphScopeNew/lib/python3.9/site-packages/gscoordinator/coordinator.py", line 316, in _RunStep
    head, bodies = self._operation_executor.run_on_analytical_engine(
  File "/Users/zhumeng/anaconda3/envs/graphScopeNew/lib/python3.9/site-packages/gscoordinator/monitor.py", line 191, in runOnAnalyticalEngineWarp
    res = func(instance, dag_def, dag_bodies, loader_op_bodies)
  File "/Users/zhumeng/anaconda3/envs/graphScopeNew/lib/python3.9/site-packages/gscoordinator/op_executor.py", line 172, in run_on_analytical_engine
    dag_def, dag_bodies = self.pre_process(dag_def, dag_bodies, loader_op_bodies)
  File "/Users/zhumeng/anaconda3/envs/graphScopeNew/lib/python3.9/site-packages/gscoordinator/op_executor.py", line 147, in pre_process
    op, _, _ = self._maybe_compile_app(op)
  File "/Users/zhumeng/anaconda3/envs/graphScopeNew/lib/python3.9/site-packages/gscoordinator/op_executor.py", line 263, in _maybe_compile_app
    compiled_path = self._compile_lib_and_distribute(
  File "/Users/zhumeng/anaconda3/envs/graphScopeNew/lib/python3.9/site-packages/gscoordinator/op_executor.py", line 403, in _compile_lib_and_distribute
    lib_path, java_jar_path, java_ffi_path, app_type = compile_func(
  File "/Users/zhumeng/anaconda3/envs/graphScopeNew/lib/python3.9/site-packages/gscoordinator/utils.py", line 572, in compile_app
    lib_path = compile_library(commands, library_dir, library_name, launcher)
  File "/Users/zhumeng/anaconda3/envs/graphScopeNew/lib/python3.9/site-packages/gscoordinator/utils.py", line 336, in compile_library
    return _compile_on_local(commands, workdir, output_name)
  File "/Users/zhumeng/anaconda3/envs/graphScopeNew/lib/python3.9/site-packages/gscoordinator/utils.py", line 389, in _compile_on_local
    raise CompilationError(
graphscope.framework.errors.CompilationError: Failed to compile 1a967046623a450bed03c8b6f754f261080f656331f81e5784248a80158b93da on platform system: Darwin
machine: arm64
platform: macOS-12.5-arm64-arm-64bit
uname: uname_result(system='Darwin', node='zhumengdeMacBook-Pro.local', release='21.6.0', version='Darwin Kernel Version 21.6.0: Sat Jun 18 17:07:22 PDT 2022; root:xnu-8020.140.41~1/RELEASE_ARM64_T6000', machine='arm64')
kernel_ver: Darwin Kernel Version 21.6.0: Sat Jun 18 17:07:22 PDT 2022; root:xnu-8020.140.41~1/RELEASE_ARM64_T6000
mac_ver: ('12.5', ('', '', ''), 'arm64')
gcc_ver: b'Apple clang version 14.0.0 (clang-1400.0.29.202)\nTarget: arm64-apple-darwin21.6.0\nThread model: posix\nInstalledDir: /Library/Developer/CommandLineTools/usr/bin\n'
python_ver: 3.9.18


╭───────────────────── Traceback (most recent call last) ──────────────────────╮
│ /Users/zhumeng/Documents/routeCompute/routeCompute/oneway/routeSeekPIE.py:47 │
│ in <module>                                                                  │
│                                                                              │
│   44 │   print("装载全图共耗时:", time.time() - startTime, "ms")             │
│   45 │   #---------------------自定义PIE算法模型调用---------------          │
│   46 │   # c = graphscope.degree_centrality(routeGraph, centrality_type="bot │
│ ❱ 47 │   c = graphscope.pagerank(routeGraph, delta=0.85, max_round=10)       │
│   48 │   r1 = c.to_numpy("r")                                                │
│   49 │   r2 = c.to_dataframe({"id": "v.id", "result": "r"})                  │
│   50 │   print(r1)                                                           │
│                                                                              │
│ /Users/zhumeng/anaconda3/envs/graphScopeNew/lib/python3.9/site-packages/grap │
│ hscope/framework/app.py:76 in wrapper                                        │
│                                                                              │
│    73 │   │   │   projected._base_graph = graph                              │
│    74 │   │   else:                                                          │
│    75 │   │   │   projected = graph                                          │
│ ❱  76 │   │   return func(projected, *args[1:], **kwargs)                    │
│    77 │                                                                      │
│    78 │   return wrapper                                                     │
│    79                                                                        │
│                                                                              │
│ /Users/zhumeng/anaconda3/envs/graphScopeNew/lib/python3.9/site-packages/grap │
│ hscope/framework/app.py:142 in wrapper                                       │
│                                                                              │
│   139 │   │   │   │   │   % (not_compatible_for_func.__name__, tag)          │
│   140 │   │   │   │   )                                                      │
│   141 │   │   │   else:                                                      │
│ ❱ 142 │   │   │   │   return not_compatible_for_func(*args, **kwargs)        │
│   143 │   │                                                                  │
│   144 │   │   return wrapper                                                 │
│   145                                                                        │
│                                                                              │
│ /Users/zhumeng/anaconda3/envs/graphScopeNew/lib/python3.9/site-packages/grap │
│ hscope/analytical/app/pagerank.py:60 in pagerank                             │
│                                                                              │
│    57 │   algo = "pagerank_directed" if graph.is_directed() else "pagerank"  │
│    58 │   delta = float(delta)                                               │
│    59 │   max_round = int(max_round)                                         │
│ ❱  60 │   return AppAssets(algo=algo, context="vertex_data")(graph, delta, m │
│    61                                                                        │
│    62                                                                        │
│    63 @project_to_simple                                                     │
│                                                                              │
│ /Users/zhumeng/anaconda3/envs/graphScopeNew/lib/python3.9/site-packages/grap │
│ hscope/framework/app.py:324 in __call__                                      │
│                                                                              │
│   321 │                                                                      │
│   322 │   def __call__(self, graph, *args, **kwargs):                        │
│   323 │   │   """Instantiate an App and do queries over it."""               │
│ ❱ 324 │   │   app_ = graph.session._wrapper(AppDAGNode(graph, self))         │
│   325 │   │   return app_(*args, **kwargs)                                   │
│   326                                                                        │
│   327                                                                        │
│                                                                              │
│ /Users/zhumeng/anaconda3/envs/graphScopeNew/lib/python3.9/site-packages/grap │
│ hscope/client/session.py:955 in _wrapper                                     │
│                                                                              │
│    952 │                                                                     │
│    953 │   def _wrapper(self, dag_node: DAGNode) -> Union[DAGNode, App, Cont │
│    954 │   │   if self.eager():                                              │
│ ❱  955 │   │   │   return self.run(dag_node)                                 │
│    956 │   │   return dag_node                                               │
│    957 │                                                                     │
│    958 │   def run(self, fetches):                                           │
│                                                                              │
│ /Users/zhumeng/anaconda3/envs/graphScopeNew/lib/python3.9/site-packages/grap │
│ hscope/client/session.py:991 in run                                          │
│                                                                              │
│    988 │   │   gc.collect()                                                  │
│    989 │   │                                                                 │
│    990 │   │   with self._lock:                                              │
│ ❱  991 │   │   │   return self.run_fetches(fetches)                          │
│    992 │                                                                     │
│    993 │   def run_fetches(self, fetches):                                   │
│    994 │   │   """Run operations of `fetches` without the session lock."""   │
│                                                                              │
│ /Users/zhumeng/anaconda3/envs/graphScopeNew/lib/python3.9/site-packages/grap │
│ hscope/client/session.py:1001 in run_fetches                                 │
│                                                                              │
│    998 │   │   │   raise RuntimeError("Session disconnected.")               │
│    999 │   │   fetch_handler = _FetchHandler(self.dag, fetches)              │
│   1000 │   │   try:                                                          │
│ ❱ 1001 │   │   │   response = self._grpc_client.run(fetch_handler.targets)   │
│   1002 │   │   except FatalError:                                            │
│   1003 │   │   │   self.close()                                              │
│   1004 │   │   │   raise                                                     │
│                                                                              │
│ /Users/zhumeng/anaconda3/envs/graphScopeNew/lib/python3.9/site-packages/grap │
│ hscope/client/rpc.py:98 in run                                               │
│                                                                              │
│    95 │   │   return str(self)                                               │
│    96 │                                                                      │
│    97 │   def run(self, dag_def):                                            │
│ ❱  98 │   │   return self._run_step_impl(dag_def)                            │
│    99 │                                                                      │
│   100 │   def fetch_logs(self):                                              │
│   101 │   │   if self._logs_fetching_thread is None:                         │
│                                                                              │
│ /Users/zhumeng/anaconda3/envs/graphScopeNew/lib/python3.9/site-packages/grap │
│ hscope/client/utils.py:156 in with_grpc_catch                                │
│                                                                              │
│   153 │   │   retries = 0                                                    │
│   154 │   │   while True:                                                    │
│   155 │   │   │   try:                                                       │
│ ❱ 156 │   │   │   │   return fn(*args, **kwargs)                             │
│   157 │   │   │   except grpc.RpcError as exc:                               │
│   158 │   │   │   │   code = exc.code()                                      │
│   159 │   │   │   │   max_retries = GRPC_MAX_RETRIES_BY_CODE.get(code)       │
│                                                                              │
│ /Users/zhumeng/anaconda3/envs/graphScopeNew/lib/python3.9/site-packages/grap │
│ hscope/client/rpc.py:198 in _run_step_impl                                   │
│                                                                              │
│   195 │   │   │   │   if isinstance(exc, tuple):                             │
│   196 │   │   │   │   │   raise exc[0](*exc[1:])                             │
│   197 │   │   │   │   else:                                                  │
│ ❱ 198 │   │   │   │   │   raise exc                                          │
│   199 │   │   return response                                                │
│   200 │                                                                      │
│   201 │   def create_analytical_instance(self):                              │
╰──────────────────────────────────────────────────────────────────────────────╯
CompilationError: Failed to compile 
1a967046623a450bed03c8b6f754f261080f656331f81e5784248a80158b93da on platform 
system: Darwin
machine: arm64
platform: macOS-12.5-arm64-arm-64bit
uname: uname_result(system='Darwin', node='zhumengdeMacBook-Pro.local', 
release='21.6.0', version='Darwin Kernel Version 21.6.0: Sat Jun 18 17:07:22 PDT
2022; root:xnu-8020.140.41~1/RELEASE_ARM64_T6000', machine='arm64')
kernel_ver: Darwin Kernel Version 21.6.0: Sat Jun 18 17:07:22 PDT 2022; 
root:xnu-8020.140.41~1/RELEASE_ARM64_T6000
mac_ver: ('12.5', ('', '', ''), 'arm64')
gcc_ver: b'Apple clang version 14.0.0 (clang-1400.0.29.202)\nTarget: 
arm64-apple-darwin21.6.0\nThread model: posix\nInstalledDir: 
/Library/Developer/CommandLineTools/usr/bin\n'
python_ver: 3.9.18

**Reply by Zhumeng420** [reply]
我的代码如下：使用第45行可以正常输出结果，但使用第46行会抛出上诉编译错误
<img width="768" alt="截屏2023-12-30 12 25 50" src="https://github.com/alibaba/GraphScope/assets/76473360/f3a3b7db-d1ca-4395-a3f8-b3aa1caa82e5">

**Reply by siyuan0322** [reply, marked as answer]
FileNotFoundError: [Errno 2] No such file or directory: 'cmake'

需要装一下 cmake 和 g++/clang++
